Maintain the same Stripe Customer across multiple cards

brchristian opened the pull request:
The initial commit creates a (currently failing) spec to test whether multiple cards used by the same Solidus user produce the same Stripe customer (i.e., the same `gateway_customer_profile_id`).

This is the beginning of progress to address https://github.com/solidusio/solidus_stripe/issues/26. The following commits resolve the issue.

brchristian commented:
I've created a fix that addresses #26, and the specs should now be green.

It's a bit of a kludge, but I am hoping we can refactor anyway once we move away from ActiveMerchant.

This should solve #26.

brchristian commented:
Specs are passing for both V2 and V3 with Elements! 🎉 

I have never used V3 with Intents before and am unfamiliar with how it works, so I am not sure what to do there. @spaghetticode what do you think?

brchristian commented:
Now that https://github.com/solidusio/solidus_stripe/pull/78 has been merged, we can make these new specs a bit cleaner / more concise. I have rebased to current `master` and added a new commit https://github.com/solidusio/solidus_stripe/pull/77/commits/39fbb6926c893cd50b4384c819feb5d84e99b341 to use the `fill_in_card` helper.

spaghetticode commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -0,0 +1,178 @@
+# frozen_string_literal: true
```
A lot of code from this file looks to be the same as the one from [stripe_checkout.spec](https://github.com/solidusio/solidus_stripe/blob/master/spec/features/stripe_checkout_spec.rb). For example, the `before` block is the same. I think it would be nice if we could avoid so much duplication. What about using shared examples, if feasible, or/and shared helper methods so we can DRY-up the files?

brchristian replied:
Not a problem to DRY up the `before` block especially. In https://github.com/solidusio/solidus_stripe/pull/78 I was asked to put a helper into `lib/solidus_stripe/testing_support`. Do you want something there as well, or would you want it to live in the `spec/` path?

brchristian replied:
Update! Was able to remove 172 lines of code from the spec by more aggressively using the `before` block and `shared_examples`. :)

brchristian commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -207,4 +207,163 @@
+
+      stripe_customer_cards = Stripe::PaymentMethod.list({ customer: stripe_customer.id, type: 'card' })
+      expect(stripe_customer_cards.count).to eq(2)
+      expect(stripe_customer_cards.data.map { |card| card.id }.sort).to eq(u.orders.map { |o| o.payments.valid.first.source.gateway_payment_profile_id }.uniq.sort)
```
👍 Did not know about `match_array`. That's perfect!

brchristian commented on Gemfile:
```diff
@@ -23,12 +23,11 @@ else
   gem "pry-rails"
   gem "ffaker"
   gem "rails-controller-testing"
+  gem 'stripe'
```
👍

brchristian commented on app/models/spree/payment_method/stripe_credit_card.rb:
```diff
@@ -119,6 +119,12 @@ def create_profile(payment)
           creditcard = source
         end
 
+        user_payment_sources = payment&.source&.user&.wallet&.wallet_payment_sources
```
👍 Good idea. I am removing the first two safe navigation operators.

brchristian commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -120,8 +120,8 @@
       cc = u.wallet.wallet_payment_sources.first.payment_source
-      expect(cc.gateway_customer_profile_id.start_with?('cus_')).to be true
-      expect(cc.gateway_payment_profile_id.start_with?('card_')).to be true
+      expect(cc.gateway_customer_profile_id).to start_with 'cus_'
```
👍

brchristian commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -207,4 +207,163 @@
+      click_button "Place Order"
+      expect(page).to have_content("Your order has been processed successfully")
+
+      u = Spree::User.find_by_email("mary@example.com")
```
👍

brchristian commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -130,6 +141,17 @@
+      u = Spree::User.find_by_email("mary@example.com")
+      expect(u.wallet.wallet_payment_sources.size).to eq(1)
+
+      cc = u.credit_cards.first
```
👍 re variable names.

Re repetition, looking more closely I think we can simply drop these expectations, as the important ones are the ones at the end of in this scenario. These particular ones were functioning mostly as a sanity check.

brchristian commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -207,4 +207,163 @@
+      expect(page).to have_current_path("/checkout/confirm")
+      u = Spree::User.find_by_email("mary@example.com")
+      expect(u.credit_cards.size).to eq(2)
+      expect(u.credit_cards.last.gateway_customer_profile_id).to start_with 'cus_'
```
👍

brchristian commented on app/models/spree/payment_method/stripe_credit_card.rb:
```diff
@@ -130,8 +136,8 @@ def create_profile(payment)
               cc_type: payment.source.cc_type,
-              gateway_customer_profile_id: response.params['id'],
-              gateway_payment_profile_id: response.params['default_source'] || response.params['default_card']
+              gateway_customer_profile_id: options[:customer] ? response.params['customer'] : response.params['id'],
```
In general I would agree with you, but I think in context with the line that follows, it's a bit more complicated. As you can see, both ternaries use the same setup with `options[:customer] ? `.

So I think there's an argument for keeping the logic parallel between the two, even if the first one could be made shorter.

brchristian commented on app/models/spree/payment_method/stripe_credit_card.rb:
```diff
@@ -119,6 +119,12 @@ def create_profile(payment)
 
+        user_payment_sources = payment&.source&.user&.wallet&.wallet_payment_sources
+        if user_payment_sources.present?
+          customer_id = user_payment_sources.map {|ps| ps.payment_source&.gateway_customer_profile_id }.compact.last
```
👍

brchristian commented:
@spaghetticode Thanks for the code review!

I squashed into two commits (specs + change) and made all of your suggested changes with two exceptions where I've followed up and which you can see above. LMK on those last two and then we should be just about there!

brchristian commented:
Update: I returned to the specs and was able to dedupe 172 lines by following your suggestion. :)

That just leaves the last question about `stripe_credit_card.rb`. My personal opinion is to leave it as is, for the reasons I explained on the diff, but if you think it’s better to make that change, LMK and I defer to you.

igorbp commented:
Hi, I just want to raise a question here :

It's possible to delete a customer in the stripe dashboard, and I couldn't find a way to recreate it with the same ID. When it happens it will break this code. I'm not sure if this should be considered or not.

@brchristian @spaghetticode

spaghetticode commented:
@igorbp thanks for raising this question 💯 

I don't know if it's very common for admins to delete accounts on the stripe dashboards, but I think it's something that we should consider fixing although not necessarily in the first implementation of this feature.

igorbp commented:
@spaghetticode I agree with you. 

Also, I'm aware of #81 and #82. But I think we could move this forward before start the change to the official stripe gem.

brchristian commented:
Hi @igorbp I second @spaghetticode in saying this is a very good point. Though we might want to visit that in a future PR.

@spaghetticode I agree with @igorbp that it might make sense to merge this before we look at either #81 or #82 which introduce additional changes. Ideally I would like to add some tests as well, because there is certain behavior that differs between Active Merchant and the Stripe Gem – for instance, whether to return `nil` or throw an exception when card data is bad – which is not part of current specs (or exists in Active Merchant specs but should probably go here as well, especially if we move away from Active Merchant). But I think all of that we can deal with later when thinking about #81 and/or #82.

In the meantime, however, anything else that I can do to help get this over the finish line?

kennyadsl commented on Gemfile:
```diff
@@ -27,6 +27,7 @@ group :development, :test do
   gem "pry-rails"
   gem "ffaker"
   gem "rails-controller-testing"
+  gem "stripe"
```
Why do we need to introduce this development dependency? Is it just for testing purposes?

brchristian replied:
@kennyadsl Exactly, I used the Stripe gem in the specs as a source for "ground-truth" customer and payment info.

spaghetticode reviewed (commented):
@brchristian thanks for addressing the issues ❤️  

I think the change is good, I left some more comments on the code. 

One more thing, I would expand the commit message for [5d43f](https://github.com/solidusio/solidus_stripe/pull/77/commits/5d43fed80b8832a19aa1380d1d2077794cea3b69) as it's currently not explaining the reason of the change. A good description may help people debugging and understanding what's going on in case of unexpected behavior.

spaghetticode commented on spec/features/stripe_customer_spec.rb:
```diff
@@ -0,0 +1,176 @@
+# frozen_string_literal: true
+
```
The first 100 lines here are still almost identical to the ones in stripe_checkout.spec. I think that all this duplication may become an issue in the future, as changes in one file will need to be ported to the other, for example if something changes on Solidus FE flow.

brchristian replied:
Good point. I’ve moved this to a `checkout_helper`, into a method called `initialize_checkout`. This deduplicates almost all of the code that was shared between `stripe_checkout_spec` and  `stripe_customer_spec`.

brchristian commented on app/models/spree/payment_method/stripe_credit_card.rb:
```diff
@@ -119,6 +119,14 @@ def create_profile(payment)
         end
 
+        user_stripe_payment_sources = payment.source.user&.wallet&.wallet_payment_sources&.select do |wps|
+          wps.payment_source.payment_method.type == 'Spree::PaymentMethod::StripeCreditCard'
```
Great idea! I’ve made the change and force-pushed.

brchristian commented:
@spaghetticode I've gone ahead and followed all of your suggestions! I've...

- Updated `self.class.name` in place of the hardcoded string
- Deduplicated the spec code into a `checkout_helper` in the `testing_support` folder as in https://github.com/solidusio/solidus_stripe/pull/78
- Expanded the commit message (which now has the hash 74f5abb70d3fe1df43090b104902adc6511c1e10).
- Rebased to the current `master`
- Force-pushed

...and CI is green!

brchristian commented on lib/solidus_stripe/testing_support/checkout_helper.rb:
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module SolidusCheckoutHelper
```
@spaghetticode Sounds good. I've renamed to `StripeCheckoutHelper` and `stripe_checkout_helper.rb` and have force-pushed the latest!

brchristian commented:
@spaghetticode What else is needed to push this over the finish line?

spaghetticode commented:
@brchristian sorry for the late answer. I think the changes look good, but the CI is failing. The failure is not related to your changes, but I think it's best to wait for the green. I'll keep an eye on the fix (either in Solidus or Solidus Support) and ping you back if any rebase/change is needed, thanks!

brchristian commented:
@spaghetticode Now that #99 has been merged, the specs are looking good except for one issue, that only seems to come up in Solidus v2.10 and seems unrelated to these changes – I also notice it in the failing specs for #92, for example.

Is this just a fluke, or is there more work to be done on `master` to get that passing? LMK if I can help.

spaghetticode commented:
@brchristian can you please rebase this branch with master? The postgres build is now failing because of Solidus gem incompatibilities that are solved in master, so hopefully after that it should be passing again. Thanks 🙏

kennyadsl commented:
I think we can close this one. It is covered in the new stripe version (v5). Thanks anyway!
